### PR TITLE
STOPS-14

### PR DESCRIPTION
**Pull request by julian-mathes**
This PR adds tag and merge functionality and adds our company internal githooks

### CODE REVIEW

File: .githooks/prepare-commit-msg [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
+
+# All .githooks directories originate 
+# from the .githooks directory in the devops repo
+
+COMMIT_MSG_FILE=$1
+COMMIT_SOURCE=$2
+SHA1=$3
```
Comment (MPStudyly) [inline .githooks/prepare-commit-msg:8]: Not relevant for this PR, but we should rename this to a more generic 'COMMIT_HASH' instead, as it is already being worked on surpassing SHA1 due to its flaws. Maybe we can open a ticket on our STOPS board?

File: ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py [hunk 2]
```diff
@@ -131,6 +131,31 @@ def add(self):
...
+        Fails if branch couldn't be checked out.
+
+        args:
+            * module:
+                type: dict()
+                description: Ansible basic module utilities and module arguments.
+        return:
+            * data:
+                type: set()
+                description: list of files changed in repo.
```
Comment (MPStudyly) [inline ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py]: Did you forget to update this description? It looks to me as if a whole dict is returned by this, instead of a simple set as e.g. on `status`.

File: ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py [hunk 3]
```diff
@@ -183,6 +208,36 @@ def commit(self):
...
+            * module:
+                type: dict()
+                description: Ansible basic module utilities and module arguments.
+        return:
+            * result:
+                type: dict()
+                description: returned output from git commit command and changed status
+        """
+        merging_branch = self.module.params.get("merge")
+        merge_options = self.module.params.get("merg_options")
```
Comment (MPStudyly) [inline ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py]: Is there a typo in "merg_options"?

File: ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py [hunk 4]
```diff
@@ -183,6 +208,36 @@ def commit(self):
             }
         FailingMessage(self.module, rc, command, output, error)
 
+    def merge(self):
+        """
+        Merge into current repo.
```
Comment (MPStudyly) [inline ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py]: Maybe we could make this a bit more generic by providing an option to specify the target branch instead of defaulting to the currently checked out one. The `git merge` command should support a target branch out of the box. This would avoid the need to checkout the proper branch before calling this task.

Nitpicky: It should state "Merge into current branch", as "repo" suggests this might be performing some sort of cherry picking from foreign repos.

File: ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py [hunk 5]
```diff
@@ -183,6 +208,36 @@ def commit(self):
...
+        args:
+            * module:
+                type: dict()
+                description: Ansible basic module utilities and module arguments.
+        return:
+            * result:
+                type: dict()
+                description: returned output from git commit command and changed status
+        """
+        merging_branch = self.module.params.get("merge")
```
Comment (MPStudyly) [inline ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py]: I'd rename this to "source_branch" to make it more obvious what this is. Especially if we introduce a "target_branch".

File: ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py [hunk 6]
```diff
@@ -247,6 +303,9 @@ def push(self):
 
         if push_force:
             command.append("--force")
+
+        if tag:
+            command.append("--tags")
```
Comment (MPStudyly) [inline ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py:299]: Can you check if pushing with parameter `--tags` does indeed push the tags and changes? I've already ran into instances where `--tags` did only push tags, but this might be local git config dependent. If this is not meant to push both, you can ignore this comment.

File: ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py [hunk 7]
```diff
@@ -255,3 +314,33 @@ def push(self):
...
+                type: list()
+                description: list of commands to perform git push operation.
+        return:
+            * result:
+                type: dict()
+                description: returned output from git push command and updated changed status.
+        """
+        tag_name: str = self.module.params["tag"]
+
+        command = [self.git_path, "tag", "-am", tag_name, tag_name]
```
Comment (MPStudyly) [inline ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py]: We should give the user the possibility to choose the message attached to the tag, if we create tags with messages. Otherwise we should simply create tags without messages (need to look up the flag with tells git to skip the message).

File: ansible_collections/lvrfrc87/git_acp/plugins/modules/git_acp.py [hunk 8]
```diff
@@ -277,17 +286,27 @@ def main():
...
+    elif tag:
+        result.update(git.tag())
+    else:
+        push = False
 
-        if push:
-            result.update(git.push())
+    if push:
+        result.update(git.push())
         result["changed"] = True
```
Comment (MPStudyly) [inline ansible_collections/lvrfrc87/git_acp/plugins/modules/git_acp.py]: Is this indentation correct? It seems something got scrambled up here.

File: ansible_collections/lvrfrc87/git_acp/tests/tests_own/main.yml [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  hosts: localhost
+  connection: local
+
+  collections:
+    - lvrfrc87.git_acp
+
+  tasks:
+  - name: CHECKOUT TO BRANCH-2
+    git_acp:
+      path: "$HOME/Projects/test-automatic-deploy/"
```
Comment (MPStudyly) [inline ansible_collections/lvrfrc87/git_acp/tests/tests_own/main.yml]: How does the original module perform tests? This paths are specific to your machine and will probably not work in other environments. If possible, adjust this to match other existing tests. If just relevant locally on your machine (e.g. while developing), eliminate it from the repo.

**Review by MPStudyly — CHANGES_REQUESTED** [review]
As we modified this module to our needs and it probably won't be merged back to upstream, we should consider renaming it on the Ansible level as well. We should have another call about this though.

File: ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py (hunk 2 above)
Comment (julian-mathes) [reply ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py]: This is a copy past error from the provided documentation as it is really all over the place and sometime not accurate at all.

File: ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py (hunk 3 above)
Comment (julian-mathes) [reply ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py]: Yes, yes there is!

File: ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py (hunk 4 above)
Comment (julian-mathes) [reply ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py]: if this is possible to specify the target branch for every git command we could do that. Otherwise I would leave it as it is since the module currently always checks out the branch if one is specified in the ansible playbook. 

You are right about the repo-branch confusion! Noted and changed!

File: ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py (hunk 6 above)
Comment (julian-mathes) [reply ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py:299]: `--tags` does only push the tags. I don't think that's a problem tho since tasks in my understanding should be as "small" as possible. So you shouldn't really push changes AND tags in the same task.

Maybe there's a way to specify that certain parameter can't be used if certain others are given. That would throw an error in the ansible playbook and would force us to keep those tasks as simple as possible

File: ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py (hunk 7 above)
Comment (julian-mathes) [reply ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py]: Can do, will do!

File: ansible_collections/lvrfrc87/git_acp/plugins/modules/git_acp.py (hunk 8 above)
Comment (julian-mathes) [reply ansible_collections/lvrfrc87/git_acp/plugins/modules/git_acp.py]: that's intended

File: ansible_collections/lvrfrc87/git_acp/tests/tests_own/main.yml (hunk 9 above)
Comment (julian-mathes) [reply ansible_collections/lvrfrc87/git_acp/tests/tests_own/main.yml]: Oh yeah complete forgot about that file.

File: ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py (hunk 2 above)
Comment (MPStudyly) [reply ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py]: Correcting our docs is enough, no need to fix forked docs.

File: ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py (hunk 4 above)
Comment (MPStudyly) [reply ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py]: Parameter thing solved, won't be changed.

Docs fix still open.

File: ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py (hunk 6 above)
Comment (MPStudyly) [reply ansible_collections/lvrfrc87/git_acp/plugins/module_utils/git_actions.py:299]: This works as intended, mutual exclusion of parameters will be implemented though.

File: ansible_collections/lvrfrc87/git_acp/plugins/modules/git_acp.py (hunk 8 above)
Comment (MPStudyly) [reply ansible_collections/lvrfrc87/git_acp/plugins/modules/git_acp.py]: Okay! Just make sure "changed" is always set properly, thus set to "true" only when a change is actually occurred.

File: README.md [hunk 10]
```diff
@@ -1,26 +1,26 @@
-# git-acp-ansible
+# git-mactp-ansible
 
-`git_acp` is an Ansible module for `git add`, `git commit`, `git push` and `git pull` operations on local or remote (https/ssh) git repo. The module will interact with the local shell execution environment, so certain commands such as setting a new git URL will edit the local `.git/config`.
+`git_mactp` is an Ansible module for `git add`, `git commit`, `git push` and `git pull` operations on local or remote (https/ssh) git repo. The module will interact with the local shell execution environment, so certain commands such as setting a new git URL will edit the local `.git/config`.
```
Comment (MPStudyly) [inline README.md]: For completion, we should add the new features to this description.

File: README.md [hunk 11]
```diff
@@ -220,7 +220,7 @@ options:
         "git_push": {
             "changed": true,
             "error": "",
-            "output": "remote: Resolving deltas:   0% (0/1)        \rremote: Resolving deltas: 100% (1/1)        \rremote: Resolving deltas: 100% (1/1), completed with 1 local object.        \nTo https://github.com/lvrfrc87/git-acp-test.git\n   3ba9041..4596d9d  master -> master\n"
+            "output": "remote: Resolving deltas:   0% (0/1)        \rremote: Resolving deltas: 100% (1/1)        \rremote: Resolving deltas: 100% (1/1), completed with 1 local object.        \nTo https://github.com/studyly/git-acp-test.git\n   3ba9041..4596d9d  master -> master\n"
```
Comment (MPStudyly) [inline README.md]: This looks like a copy-paste error and should be ".../Studyly-GmbH/..." instead.

We introduced a slight naming inconsistency by calling the collection "studyly" while the repository is owned by "Studyly-GmbH". Though I think the simple name for the collection was a good choice.

File: ansible_collections/lvrfrc87/git_acp/lvrfrc87-git_acp-1.1.0.tar.gz
Comment (MPStudyly) [inline ansible_collections/lvrfrc87/git_acp/lvrfrc87-git_acp-1.1.0.tar.gz:1]: We should think about deleting these files. We'd have to do a cleanup afterwards to properly get rid of them, but having them gone from up to date versions might be a good first step.

File: ansible_collections/studyly/git_mactp/README.md [hunk 12]
```diff
@@ -2,11 +2,11 @@ Module documentation can be found [here](../../../README.md)
 
 ### HOW INSTALL COLLECTION
 
-- Locally using `tar` file: `ansible-galaxy collection install lvrfrc87-git_acp-2.1.0.tar.gz`
+- Locally using `tar` file: `ansible-galaxy collection install studyly-git_mactp-2.1.0.tar.gz`
 
-- From GitHub: `ansible-galaxy collection install git+https://github.com/lvrfrc87/git-acp-ansible.git#ansible_collections/lvrfrc87/git_acp,master`
+- From GitHub: `ansible-galaxy collection install git+https://github.com/lvrfrc87/git-acp-ansible.git#ansible_collections/lvrfrc87/git_mactp,master`
```
Comment (MPStudyly) [inline ansible_collections/studyly/git_mactp/README.md]: Rename is missing.

File: ansible_collections/studyly/git_mactp/README.md [hunk 13]
```diff
@@ -2,11 +2,11 @@ Module documentation can be found [here](../../../README.md)
...
 
-- From [Ansible Galaxy](https://galaxy.ansible.com/): `ansible-galaxy collection install lvrfrc87.git_acp`
+- From [Ansible Galaxy](https://galaxy.ansible.com/): `ansible-galaxy collection install lvrfrc87.git_mactp`
```
Comment (MPStudyly) [inline ansible_collections/studyly/git_mactp/README.md]: Rename is missing.

File: ansible_collections/studyly/git_mactp/galaxy.yml [hunk 14]
```diff
@@ -1,5 +1,5 @@
-namespace: lvrfrc87
-name: git_acp
+namespace: studyly
+name: git_mactp
 version: 2.1.0
```
Comment (MPStudyly) [inline ansible_collections/studyly/git_mactp/galaxy.yml]: Do we want to change the version? As we added new functions, but didn't break original stuff (?), we could go to 2.2.0. If changed, don't forget to adjust other mentions of if in the docs.

File: ansible_collections/studyly/git_mactp/plugins/module_utils/git_actions.py [hunk 15]
```diff
@@ -131,6 +131,31 @@ def add(self):
...
+        Checkout branch
+        Fails if branch couldn't be checked out.
+
+        args:
+            * module:
+                type: dict()
+                description: Ansible basic module utilities and module arguments.
+        return:
+            * data:
+                type: set()
```
Comment (MPStudyly) [inline ansible_collections/studyly/git_mactp/plugins/module_utils/git_actions.py:145]: Type should be dict().

File: ansible_collections/studyly/git_mactp/plugins/module_utils/git_actions.py [hunk 16]
```diff
@@ -172,8 +197,8 @@ def commit(self):
                 type: dict()
                 description: returned output from git commit command and changed status
         """
-        comment = self.module.params["comment"]
-        command = [self.git_path, "commit", "-m", comment]
+        message = self.module.params["message"]
+        command = [self.git_path, "commit", "-m", message]
```
Comment (MPStudyly) [inline ansible_collections/studyly/git_mactp/plugins/module_utils/git_actions.py:201]: Out of curiosity, why the change?

File: ansible_collections/studyly/git_mactp/plugins/module_utils/git_actions.py [hunk 17]
```diff
@@ -183,20 +208,44 @@ def commit(self):
...
+
+        if merge_options:
+            for opt in merge_options:
+                command.insert(2, opt)
+
+        rc, output, error = self.module.run_command(command, cwd=self.path)
+
+        if rc == 0:
+            return {
+                "git_merge": {"output": output, "error": error, "changed": True}
```
Comment (MPStudyly) [inline ansible_collections/studyly/git_mactp/plugins/module_utils/git_actions.py:237]: As discussed, a merge might not actually update a branch. We should somehow evaluate the output. The simplest I can think of would be to check if the output contained something like "everything up to date" (or whatever it was exactly). To do this right, we'd have to ensure the locale on the machine where this is run is en_US or en_GB, but this seems to be a limitation of the module already, thus I'd be okay with simply assuming the locale.

File: ansible_collections/studyly/git_mactp/plugins/module_utils/git_actions.py [hunk 18]
```diff
@@ -183,20 +208,44 @@ def commit(self):
...
         args:
-            * url:
-                type: str()
-                description: git url of the git repo.
-            * branch:
-                type: str()
-                description: git branch of the git repo.
-            * pull_options:
-                type: list()
-                description: pull options added to the pull command.
```
Comment (MPStudyly) [inline ansible_collections/studyly/git_mactp/plugins/module_utils/git_actions.py:199]: Why were these removed? The code below suggests these parameters still exist.

File: ansible_collections/studyly/git_mactp/plugins/module_utils/git_actions.py [hunk 19]
```diff
@@ -224,12 +273,9 @@ def push(self):
         Push changes to remote repo.
 
         args:
-            * path:
-                type: path
-                description: git repo local path.
-            * cmd_push:
-                type: list()
-                description: list of commands to perform git push operation.
```
Comment (MPStudyly) [inline ansible_collections/studyly/git_mactp/plugins/module_utils/git_actions.py:232]: Were these obsolete already? The code below shows no signs of removal of related stuff.

Maybe you could add docs about the actual valid params here?

File: ansible_collections/studyly/git_mactp/plugins/module_utils/git_actions.py [hunk 20]
```diff
@@ -255,3 +305,36 @@ def push(self):
                 "git_push": {"output": str(output), "error": str(error), "changed": True}
```
Comment (MPStudyly) [inline ansible_collections/studyly/git_mactp/plugins/module_utils/git_actions.py:305]: Same as on "merge", we should ensure something actually changed.

File: ansible_collections/studyly/git_mactp/plugins/module_utils/git_actions.py [hunk 21]
```diff
@@ -255,3 +305,36 @@ def push(self):
...
+        # - if message was not given, prompt for message
+        # - if message was given but is empty, don't use any
+        # - if message was given and is not empty, use this as message
+
+        # Push result is returned in stderr instead of stdout, hence vars position is inverted.
+        rc, error, output = self.module.run_command(command, cwd=self.path)
+        
+        if rc == 0:
+            return {
+                "git_tag": {"output": str(output), "error": str(error), "changed": True}
```
Comment (MPStudyly) [inline ansible_collections/studyly/git_mactp/plugins/module_utils/git_actions.py:338]: Just a note: Assuming a change is valid in this case as it is forbidden to create a tag twice, thus git will fail instead of accepting with "nothing changed".

File: ansible_collections/studyly/git_mactp/plugins/modules/git_mactp.py [hunk 22]
```diff
@@ -277,17 +287,27 @@ def main():
...
+    elif tag:
+        result.update(git.tag())
+    else:
+        push = False
 
-        if push:
-            result.update(git.push())
+    if push:
+        result.update(git.push())
         result["changed"] = True
```
Comment (MPStudyly) [inline ansible_collections/studyly/git_mactp/plugins/modules/git_mactp.py]: I know we talked and decided on this being fine, but as I saw the actual code of the "push" function, why is it necessary to overwrite `changed` here? The function itself, if successful, will always return `changed == true` in its current shape. If refactored as suggested there, it might not. But it still always should return the proper value. Maybe I'm just overlooking something here.

File: ansible_collections/studyly/git_mactp/tests/install/ansible_collections/lvrfrc87/git_acp/FILES.json
Comment (MPStudyly) [inline ansible_collections/studyly/git_mactp/tests/install/ansible_collections/lvrfrc87/git_acp/FILES.json:1]: I guess we have to check what this file is actually for. Due to it containing SHA sums, it might actually be out of date as of forking the repo.

File: ansible_collections/studyly/git_mactp/tests/install/ansible_collections/lvrfrc87/git_acp/MANIFEST.json
Comment (MPStudyly) [inline ansible_collections/studyly/git_mactp/tests/install/ansible_collections/lvrfrc87/git_acp/MANIFEST.json:1]: I guess we want to change this info as well? As I just noticed, it is GPL licensed, so we are actually forced to make this publicly available ^^" Should we change the author though?

File: ansible_collections/studyly/git_mactp/tests/install/ansible_collections/lvrfrc87/git_acp/README.md
Comment (MPStudyly) [inline ansible_collections/studyly/git_mactp/tests/install/ansible_collections/lvrfrc87/git_acp/README.md:1]: Is this a different README than the other one? The versions do not match, while the text seems to do so.

File: pyproject.toml
Comment (MPStudyly) [inline pyproject.toml:1]: IF we change authorship and version, we should do so here as well.

**Review by MPStudyly — CHANGES_REQUESTED** [review]

File: ansible_collections/lvrfrc87/git_acp/lvrfrc87-git_acp-1.1.0.tar.gz
Comment (julian-mathes) [reply ansible_collections/lvrfrc87/git_acp/lvrfrc87-git_acp-1.1.0.tar.gz:1]: That should already be deleted!

File: ansible_collections/studyly/git_mactp/tests/install/ansible_collections/lvrfrc87/git_acp/FILES.json
Comment (julian-mathes) [reply ansible_collections/studyly/git_mactp/tests/install/ansible_collections/lvrfrc87/git_acp/FILES.json:1]: probably all of his tests are out of date. should we delete them?

File: ansible_collections/studyly/git_mactp/tests/install/ansible_collections/lvrfrc87/git_acp/README.md
Comment (julian-mathes) [reply ansible_collections/studyly/git_mactp/tests/install/ansible_collections/lvrfrc87/git_acp/README.md:1]: The version number is different. That might be resolved if we remove his tests as they are out of date

File: ansible_collections/studyly/git_mactp/tests/install/ansible_collections/lvrfrc87/git_acp/FILES.json
Comment (MPStudyly) [reply ansible_collections/studyly/git_mactp/tests/install/ansible_collections/lvrfrc87/git_acp/FILES.json:1]: Is this related to tests? If the tests are out of date only in regards to our changes, we should keep them and reactivate them at some point.

File: ansible_collections/studyly/git_mactp/tests/install/ansible_collections/lvrfrc87/git_acp/README.md
Comment (MPStudyly) [reply ansible_collections/studyly/git_mactp/tests/install/ansible_collections/lvrfrc87/git_acp/README.md:1]: Okay, so we need to decide on that. This can be resolved after being read.

File: CHANGELOG.rst [hunk 23]
```diff
@@ -1,3 +1,9 @@
+# Ver. 3.0.0
+- Changes to the original repo to allow general usage
```
Comment (MPStudyly) [inline CHANGELOG.rst]: Nitpicky, but if you go over it again, please change this to imperative ;)

File: README.md [hunk 24]
```diff
@@ -220,7 +230,7 @@ options:
         "git_push": {
             "changed": true,
             "error": "",
-            "output": "remote: Resolving deltas:   0% (0/1)        \rremote: Resolving deltas: 100% (1/1)        \rremote: Resolving deltas: 100% (1/1), completed with 1 local object.        \nTo https://github.com/studyly/git-acp-test.git\n   3ba9041..4596d9d  master -> master\n"
+            "output": "remote: Resolving deltas:   0% (0/1)        \rremote: Resolving deltas: 100% (1/1)        \rremote: Resolving deltas: 100% (1/1), completed with 1 local object.        \nTo https://github.com/lvrfrc87/git-acp-test.git\n   3ba9041..4596d9d  master -> master\n"
```
Comment (MPStudyly) [inline README.md]: Renamed by mistake?

**Review by MPStudyly — CHANGES_REQUESTED** [review]

File: README.md (hunk 24 above)
Comment (julian-mathes) [reply README.md]: No this was intentional as it referes to the original test repo

File: ansible_collections/studyly/git_mactp/plugins/modules/git_mactp.py (hunk 22 above)
Comment (MPStudyly) [reply ansible_collections/studyly/git_mactp/plugins/modules/git_mactp.py]: As we found out, this is actually the "overall" result. At this point in the code it shouldn't be possible to not have performed any changes, as the whole mactp module can just perform writing operations.

**Review by MPStudyly — APPROVED** [review]
